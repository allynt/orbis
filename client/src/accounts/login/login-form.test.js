import React from 'react';

import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import LoginForm from './login-form.component';

const EMAIL_PLACEHOLDER_TEXT = /email/i;
const PASSWORD_PLACEHOLDER_TEXT = /password/i;
const LOGIN_BUTTON_TEXT = 'Login';
const EMAIL_TEXT = 'test@test.com';
const PASSWORD_TEXT = 'testpassword';
const SIGN_UP = /sign\sup/i;
const KEEP_LOGGED_IN = /keep\sme\slogged\sin/i;
const WORK_EMAIL = /work\semail\saddress/i;
const I_AGREE_TEXT = 'I agree with';

/**
 * @param {Partial<Pick<import('./login-form.component').LoginProps,
 *   'isLoading'
 *   | 'activateAccount'
 *   | 'match'
 *   | 'serverErrors'
 *   | 'user'
 *   | 'minimalUser'>>} [props]
 */
const renderComponent = props => {
  const activateAccount = jest.fn();
  const login = jest.fn();
  const minimalUser = props?.minimalUser ? props.minimalUser : { accepted_terms: false }

  const utils = render(
    // @ts-ignore
    <LoginForm
      passwordMinLength={2}
      passwordMaxLength={255}
      activateAccount={activateAccount}
      login={login}
      minimalUser={minimalUser}
      {...props}
    />,
    {
      wrapper: ({ children }) => (
        <Router history={createMemoryHistory({ initialEntries: ['/'] })}>
          {children}
        </Router>
      ),
    },
  );
  return { ...utils, activateAccount, login };
};

describe('Login Form Component', () => {
  it('should render a form', () => {
    const { getByText, getByRole, getByLabelText } = renderComponent();

    expect(
      getByRole('textbox', { name: EMAIL_PLACEHOLDER_TEXT }),
    ).toBeInTheDocument();
    expect(getByLabelText(PASSWORD_PLACEHOLDER_TEXT)).toBeInTheDocument();
    expect(getByText(I_AGREE_TEXT)).toBeInTheDocument();
    expect(
      getByRole('button', { name: LOGIN_BUTTON_TEXT }),
    ).toBeInTheDocument();
  });

  it('should disable `Login` button when form is invalid', () => {
    const { getByRole } = renderComponent();

    expect(getByRole('button', { name: LOGIN_BUTTON_TEXT })).toBeDisabled();
  });

  it('should enable `Login` button when form is valid', async () => {
    const { getByText, getByRole, getByLabelText } = renderComponent();

    const loginButton = getByRole('button', { name: LOGIN_BUTTON_TEXT });

    expect(loginButton).toBeDisabled();

    userEvent.type(
      getByRole('textbox', { name: EMAIL_PLACEHOLDER_TEXT }),
      EMAIL_TEXT,
    );

    userEvent.type(getByLabelText(PASSWORD_PLACEHOLDER_TEXT), PASSWORD_TEXT);

    userEvent.click(getByText(I_AGREE_TEXT));

    await waitFor(() => expect(loginButton).not.toHaveAttribute('disabled'));
  });

  it('should not call `login` function when form is invalid and `Login` button clicked', async () => {
    const { getByRole, login } = renderComponent();

    const loginButton = getByRole('button', { name: LOGIN_BUTTON_TEXT });

    userEvent.type(
      getByRole('textbox', { name: EMAIL_PLACEHOLDER_TEXT }),
      EMAIL_TEXT,
    );

    userEvent.tab();

    userEvent.click(loginButton);

    expect(login).not.toHaveBeenCalled();
  });

  it('should call `login` function when form is valid and `Login` button clicked', async () => {
    const { getByText, getByRole, getByLabelText, login } = renderComponent();

    userEvent.type(
      getByRole('textbox', { name: EMAIL_PLACEHOLDER_TEXT }),
      EMAIL_TEXT,
    );
    userEvent.type(getByLabelText(PASSWORD_PLACEHOLDER_TEXT), PASSWORD_TEXT);

    userEvent.click(getByText(I_AGREE_TEXT));

    userEvent.click(getByRole('button', { name: LOGIN_BUTTON_TEXT }));

    await waitFor(() =>
      expect(login).toHaveBeenCalledWith({
        email: EMAIL_TEXT,
        password: PASSWORD_TEXT,
        termsAgreed: true,
      }),
    );
  });

  it('should display error well if login is unsuccessful', () => {
    const serverErrors = ['Test Error 1', 'Test Error 2', 'Test Error 3'];

    const { getByTestId } = renderComponent({ serverErrors });

    expect(getByTestId('error-well')).toBeInTheDocument();
  });

  it('calls activateAccount if the supplied user is not verified and the url has a key', () => {
    const { activateAccount } = renderComponent({
      // @ts-ignore
      user: { is_verified: false },
      // @ts-ignore
      match: { params: { key: '123' } },
    });
    expect(activateAccount).toBeCalledWith({ key: '123' });
  });

  it('shows a loading spinner if loading', () => {
    const { getByRole } = renderComponent({ isLoading: true });
    expect(getByRole('alert')).toBeInTheDocument();
  });

  it('only shows `Terms and Conditions` checkbox if user has not agreed already', () => {
    const { getByRole } = renderComponent();
    
    expect(getByRole('checkbox')).toBeInTheDocument();
  })

  it('does not show `Terms and Conditions` checkbox if user has agreed already', () => {
    const { queryByRole } = renderComponent({ minimalUser: { accepted_terms: true } });

    expect(queryByRole('checkbox')).not.toBeInTheDocument();
  })

  describe('Customer Registration Flow', () => {
    /** @type {Partial<User>} */
    const user = { registration_stage: 'CUSTOMER' };

    it('does not show the sign up link', () => {
      const { queryByRole } = renderComponent({ user });
      expect(queryByRole('link', { name: SIGN_UP })).not.toBeInTheDocument();
    });

    it('does not show keep me logged in', () => {
      const { queryByRole } = renderComponent({ user });
      expect(
        queryByRole('checkbox', { name: KEEP_LOGGED_IN }),
      ).not.toBeInTheDocument();
    });

    it('has the work email address label', () => {
      const { getByRole } = renderComponent({ user });
      expect(getByRole('textbox', { name: WORK_EMAIL })).toBeInTheDocument();
    });
  });
});
