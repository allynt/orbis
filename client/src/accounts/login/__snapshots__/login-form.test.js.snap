// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Component should render a form 1`] = `
<div>
  <form
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    novalidate=""
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-focused Mui-focused Mui-required Mui-required"
          data-shrink="true"
          for="email"
          id="email-label"
        >
          Email
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root makeStyles-root-1 makeStyles-root-6 MuiInput-underline makeStyles-underline-3 makeStyles-underline-7 MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input makeStyles-input-4 makeStyles-input-8"
            id="email"
            name="email"
            required=""
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root makeStyles-root-1 makeStyles-root-14 MuiInput-underline makeStyles-underline-3 makeStyles-underline-15 MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input makeStyles-input-4 makeStyles-input-16 MuiInputBase-inputAdornedEnd"
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              aria-label="Password Visibility Toggle"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-root-12 makeStyles-root-13"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19.81 4.188A.654.654 0 0019.348 4a.664.664 0 00-.462.188l-2.222 2.178A14.61 14.61 0 0012 5.604c-7.455 0-12 5.556-12 6.417 0 .584 2.097 3.322 5.768 5.027l-1.938 1.9a.641.641 0 00-.205.458.63.63 0 00.192.462.656.656 0 00.472.188.665.665 0 00.466-.201L19.81 5.095a.64.64 0 00.192-.454.63.63 0 00-.192-.453zM7.637 12.02a4.24 4.24 0 011.279-3.024A4.412 4.412 0 0112 7.743a4.327 4.327 0 012.469.775l-.791.776A3.244 3.244 0 0012 8.812a3.31 3.31 0 00-2.313.942 3.18 3.18 0 00-.96 2.268c.004.582.175 1.152.49 1.644l-.79.776a4.13 4.13 0 01-.79-2.421zm11.561-4.54l-3.12 3.059c.185.473.282.975.285 1.482a4.24 4.24 0 01-1.279 3.023A4.413 4.413 0 0112 16.3a4.395 4.395 0 01-1.511-.28L8.469 18c1.154.29 2.34.437 3.531.44 7.456 0 12-5.557 12-6.418 0-.53-1.731-2.842-4.802-4.54z"
                    fill="currentColor"
                  />
                  <path
                    d="M12 15.23c.484 0 .963-.104 1.401-.306a3.258 3.258 0 001.132-.866c.307-.367.527-.797.644-1.258a3.14 3.14 0 00.031-1.407l-3.85 3.773c.212.042.427.063.642.064z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-19 MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiBox-root MuiBox-root-20"
      >
        <a
          class="MuiTypography-root MuiLink-root makeStyles-root-21 MuiLink-underlineHover MuiTypography-colorPrimary"
          href="/accounts/password/reset"
        >
          Forgot password?
        </a>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12"
    >
      <span
        style="cursor: not-allowed;"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-root-22 MuiButton-contained makeStyles-contained-24 MuiButton-containedPrimary makeStyles-containedPrimary-25 Mui-disabled makeStyles-disabled-23 Mui-disabled"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Login
          </span>
        </button>
      </span>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Don't have an account? 
        <a
          class="MuiTypography-root MuiLink-root makeStyles-root-21 MuiLink-underlineHover MuiTypography-colorPrimary"
          href="/accounts/register"
        >
          Sign Up
        </a>
      </p>
    </div>
  </form>
</div>
`;
